File: api/filing/models/dividend_income_model.py
"""Dividend income models for dividend income management."""
from typing import List, Any
from datetime import date
from .api_base_model import ApiBaseModel


class EquityDividendModel(ApiBaseModel):
    """Model for equity dividend - maps to dividend_income table."""
    dividend_id: int | None = None
    filing_id: int | None = None
    dividend_type: str = "equity"
    currency: str = "INR"
    narration: str | None = None  # Optional - ERI doesn't provide
    amount: float = 0.0
    date_of_receipt: date | None = None  # Optional - ERI doesn't provide
    created_at: Any | None = None
    updated_at: Any | None = None


class RSUDividendModel(ApiBaseModel):
    """Model for RSU dividend - maps to equity_compensation_income table."""
    dividend_id: int | None = None  # Alias for equity_comp_income_id (for frontend consistency)
    equity_comp_income_id: int | None = None
    filing_id: int | None = None
    dividend_type: str = "rsu"
    currency: str = "INR"
    description: str | None = None
    amount: float = 0.0
    amount_received: float = 0.0
    amount_received_currency_type: str = "INR"
    date_of_receipt: date | None = None
    received_date: date | None = None
    tax_paid_foreign_currency_type: str | None = None
    tax_paid_outside_india: float = 0.0
    created_at: Any | None = None
    updated_at: Any | None = None


class DividendIncomeModel(ApiBaseModel):
    """Combined model for all dividend income types."""
    filing_id: int | None = None
    equity: List[EquityDividendModel] = []
    rsu: List[RSUDividendModel] = []

File: web/filing/models/income/dividend/dividend-income-model.ts
import { RSUDividendModel } from "./rsu-dividend-model"
import { EquityDividendModel } from "./equity-dividend-model"


export interface DividendIncomeModel {
  filingId: number
  equity?: EquityDividendModel[] | null
  rsu?: RSUDividendModel[] | null
}

// ==================== Widget Types ====================

/**
 * Master data from API response (camelCase)
 */
export interface DividendIncomeMasterData {
  currencyTypes?: Array<{ value: string; label: string }>
  dividendTypes?: Array<{ value: string; label: string }>
}

/**
 * Dividend income record (API format)
 */
export interface DividendIncomeRecord {
  dividendId?: number
  filingId?: number
  dividendType: 'equity' | 'rsu' // 'equity' for Equities/Stocks/Mutual Funds, 'rsu' for RSUs/ESOPs/ESSPs
  narration?: string // For equity type
  description?: string // For RSU type
  amount: number
  amountReceived?: number // For RSU type (alias for amount)
  dateOfReceipt: Date | string | null
  receivedDate?: Date | string | null // For RSU type (alias for dateOfReceipt)
  taxPaidOutsideIndia?: number // For RSU type
  amountReceivedCurrencyType?: string // Default: 'INR'
  taxPaidForeignCurrencyType?: string // Default: 'INR'
}

/**
 * API Response types
 */
export interface DividendIncomeResponse {
  data: DividendIncomeRecord
  message?: string
}

export interface DividendIncomeListResponse {
  data: DividendIncomeRecord[]
  message?: string
}

export interface DividendIncomeGetResponse {
  filing_id: number
  equity?: DividendIncomeRecord[]
  rsu?: DividendIncomeRecord[]
}

/**
 * API response structure (snake_case from backend)
 */
export interface DividendIncomeData {
  data: DividendIncomeGetResponse | null
  master_data?: DividendIncomeMasterData
}

export interface DividendIncomeBulkCreate {
  filingId: number
  equity?: DividendIncomeRecord[] | null // Array for equities/stocks/mutual funds
  rsu?: DividendIncomeRecord[] | null // Array for RSUs/ESOPs/ESSPs
}

/**
 * Entry interface for widget state
 */
export interface DividendEntry {
  tempId?: string // For new entries not yet saved
  dividendId?: number
  dividendType: 'equity' | 'rsu'
  // Equity fields
  narration?: string
  amount?: number
  dateOfReceipt?: Date | null
  // RSU fields
  description?: string
  amountReceived?: number
  receivedDate?: Date | null
  taxPaidOutsideIndia?: number
  amountReceivedCurrencyType?: string
  taxPaidForeignCurrencyType?: string
}

/**
 * Form data structure for DividendIncomeWidget
 */
export interface DividendIncomeFormData {
  dividendType: 'equity' | 'rsu'
  // Equity fields
  narration: string
  amount: string
  dateOfReceipt: Date | null
  // RSU fields
  description: string
  amountReceived: string
  receivedDate: Date | null
  taxPaidOutsideIndia: string
  editingIndex?: number // Index of entry being edited
  taxPaidForeignCurrencyType: string
  amountReceivedCurrencyType: string
}

/**
 * Props for DividendIncomeWidget component
 */
export interface DividendIncomeWidgetProps {
  isOpen: boolean
  onClose: () => void
  filingId: number | null
  entityId?: number | string
  initialData?: Record<string, any> | null
  onSuccess?: (action: 'created' | 'updated', data: any) => void
  errorField?: string | null
  errorMessage?: string | null  // Error message to display for the field
  fieldErrors?: Record<string, string> | null  // Multiple field errors: { fieldName: errorMessage }
}
